### ISSUE TITLE
Python 3.7 support

### ISSUE THREAD

**Issue by Harmon758** [Shizmob/pydle#76]
[Python 3.7.0](https://www.python.org/downloads/release/python-370/) has been [released](https://pythoninsider.blogspot.com/2018/06/python-3.html), and [async is now a reserved keyword in 3.7](https://docs.python.org/3.7/whatsnew/3.7.html).

**Comment by theunkn0wn1** [comment]
I have looked into what it would take to convert Pydle to be 3.7 compatible. Renaming the `pydle/async.py` module to some other name theoretically work as a stop-gap measure. 
However, as the test suite is broken, i cannot verify this is the only required change.

To quote from Shizmob 
>The asyncio branch is the future for pydle, so I don't think I'll be spending much work on making it work with Tornado 5.0+.

If `asyncio` is the future, prove useful to make direct use of the `asyncio` library rather than writing a wrapper around it, as this wrappper introduces needless complexity around features implemented directly in asyncio. It The docstring describes the function of this library is to provide a generic interface to an arbitrary back-end async library. 


The largest usage of this `async` wrapping module is the `@async.coroutine` decorator. It appears Pydle is written using language from Python 3.3 (`yield from`) and `@coroutine` to make corotines from generators definitions. 
Ex,
```py
# pydle/pydle/connection.py:Connection
    ...

    @async.coroutine
    def send(self, data):
        """ Add data to send queue. """
        self.writer.write(data)
        yield from self.writer.drain()

```
This language is largely superseded since python 3.5 with native coroutines as defined in [PEP 492](https://www.python.org/dev/peps/pep-0492/)
Unless there is a need for these coroutines to be generators, we can remove the `async` coroutine decorator entirely by using `async def` and `await` as introduced in python 3.5


The above coroutine can be rewritten as
```py
    async def send(self, data):
        """ Add data to send queue. """
        self.writer.write(data)
        await self.writer.drain()
```
If pydle needs to preserve the existing generator definitions, (I may have missed something),
`@async.coroutine` can be replaced with `@asyncio.coroutine`

As for other minor usages of this `async` wrapper, i have only noticed a small portion of the exposed methods in `async` are actually used, and can be directly implemented with calls into `asyncio`.

I am currently investigating the validity of this approach from a technical standpoint, (see [my fork's todo list](https://github.com/theunkn0wn1/pydle/issues/2) for further details on my progress.


Regardless of which approach is taken, projects will break in python 3.7 if they make use of `pydle.async` due to `async` becoming a reserved keyword.

If my approach works il open a Pull Request.

**Comment by Shizmob** [comment]
Hmm, async becoming reserved is very unfortunate. I'm not quite sure what the best approach is just yet, but I don't immediately feel like dropping the `async` wrapper entirely is it...

**Comment by prg318** [comment]
... can't you just ```sed``` to name it something else?

You don't need to drop the wrapper entirely because of a name space collision...

**Comment by theunkn0wn1** [comment]
I opted to drop the wrapper as it made debugging the failing unit tests harder due to an extra layer of abstraction. IMHO, this wrapper is a needless layer of complexity, however it is ultimately Shizmob's decision whether to simply rename the module as a workaround, or remove the wrapper entirely and make native use of Asyncio.

That said, in theory renaming the module to something other than `async` would resolve the immediate incompatibility. I have opened a PR where i took the removal approach, but can somewhat easily adapt it to simply rename the offending module should that be the desired path.

**Comment by benrob0329** [comment]
FYI, Pydle is now broken on Debian testing because of this.

**Comment by theunkn0wn1** [comment]
I have attempted to reach @Shizmob via twitter since he seems to be active there, with a little luck he had simply forgotten about Pydle and/or Github as a whole due to his lack of (public) commit activity.

Failing a response in the next few days (or perhaps after the new year, given the holiday season), I may decide to go ahead and apply this Pr to my fork and figure out how to publish it on PyPy. (S)He has been given ample time to decide if he wants to merge or reject the changes, and its becoming a blocker for my project. 

 This issue has been open for quite some time and i believe it is more than long enough to `mull over` the changes. I have been developing against the application of this PR and haven't encountered any issues, and would be willing to correct any bugs reported moving forward should i publish the fork.

**Comment by Shizmob** [comment]
Alright, so first I'd like to apologize badly. The entirety of Pydle (and most of my public open source contributions) had completely slipped off my radar due to extremely turbulent things going in other aspects of my life, and I was only reminded of its existence by the fact that [@theunkn0wn1 pinged me on Twitter about it today](https://twitter.com/0rionsEdge/status/1076367068344115200).

It wasn't responsible of me to let it wither like this having actual users, and I think the project and you all as users deserve better than that. My sincerest apologies for letting you all down.

@theunkn0wn1's PR seems fine as it stands after reviewing it, although I'd like to reintroduce `pydle.Future` and `pydle.coroutine` as existing code uses this, even if it's just light wrappers around `asyncio`.

That said, I don't feel like I'll be able to contribute to pydle meaningfully in at least the near future due to other responsibilities and projects going on, sapping the time I'd be able to invest here. I feel it would indeed be better if someone else was willing to take up the mantle, for the users' sake.
@theunkn0wn1: if you're okay with this and willing, I can add you as a maintainer to this repo and give you publishing rights on PyPI for pydle. I feel like this would be slightly preferable to a fork, but I can understand either road. Sorry for letting y'all down.

**Comment by theunkn0wn1** [comment]
Thanks for the reply! was worried you had fallen off the face of the earth for a while there! 😅 


I Will re-add the `pydle.Future` and `pydle.coroutine` objects as wrappers around their Asyncio counterparts, i was not aware they were actually in use outside Pydle itself. I will push this up later today.

I am willing to take up the mantle and continue maintaining Pydle, its a great project that just needs some TLC to keep up with the times. @Shizmob I am willing to maintain this project, fixing issues as they are raised and resolving breaking changes as they may be introduced in Python itself.
i am [`Theunkn0wn1`](https://pypi.org/user/Theunkn0wn1/) on PyPy

A fork is less desirable, simply due to the confusion it may bring, I believe it would be easier for everyone downstream if updates can be released to the existing pypy packages and/or by updating the git submodules.

**Comment by Shizmob** [comment]
Thank you, both for your hard work and your continued willingness to help out the project! I've added you as collaborator both here on GitHub and on PyPI now, and think this is a great way forward. 👍 

I've merged the hard work in #79 to `asyncio` and `master` as well now. One final note: when making a release, please tag the commit in question that bumps the release with `v{version}`. Ideally note the version number in the commit message as well. 😉

**Comment by theunkn0wn1** [comment]
Thanks, Il see to the release of `0.8.5` shortly 🙂 

I have taken note the versioning and tags, il see if i can push up a tag and set the release in motion.

My plan over the next few weeks is to look into the open tickets and see what still needs fixing, perhaps see if i can modernize the test suite.

**Comment by theunkn0wn1** [comment]
Resolved in a705380

**Comment by Harmon758** [comment]
Shouldn't this have been a v0.9 release?
Since this update included the asyncio port itself and not just Python 3.7 support.
This seems like a major release instead of a maintenance update.
There also seems to be [a lot of changes](https://github.com/Shizmob/pydle/compare/v0.8.4...v0.8.5) missing from the release notes.

**Comment by theunkn0wn1** [comment]
Looks like I did not thoroughly read the change notes myself, yes this should have been a point release as there was a larger set of changes between the two branches than just my PR. 

As for the compatibility issues, #91 #89 
The documentation issues, #94 
And the miscellaneous issues #87 #88 and #93 ,
I greatly appreciate you reporting them, but I would also appreciate the opportunity to correct them myself in one go rather than in a collection of patches.

Support for 3.5 was removed by a programming error and will be corrected after the new year.  My apologies for any inconvenience,  I will correct it as part of a larger multi-fix slated for the point-release.

For the documentation, please give me some time to see to them,most likely after the point-release.

**Comment by Harmon758** [comment]
These were separate issues and hence why I made separate issues/PRs for each.
All the PRs, including #90, which fixes support for Python 3.5, should be able to be merged as is right now and fix their corresponding issues.
